SMT2 back-end: improve support for zero-sized components

tautschnig opened the pull request:
Kani uses empty structs and unions on a regular basis, which necessitates better support for these in the SMT back-end. As SMT LIB has no provision for zero-sized bit vectors we need to filter out any such expressions.

Fixes: #7442

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
SMT2 back-end: improve support for zero-sized components

Kani uses empty structs and unions on a regular basis, which
necessitates better support for these in the SMT back-end. As SMT LIB
has no provision for zero-sized bit vectors we need to filter out any
such expressions.

Fixes: #7442

tautschnig committed:
SMT back-end: fix parsing of Z3 models involving let

Z3 produces models such as
```
((|main::1::arr!0@1#3[[1]]| (let ((a!1 (store (store (store ((as const (Array (_ BitVec 64) (_ BitVec 8)))
                                  #x00)
                                #x0000000000000003
                                #x02)
                         #x0000000000000002
                         #x01)
                  #x0000000000000004
                  #x01)))
  (mk-struct.0 #x00 #x00 a!1))))
```
(for cbmc/Array_operations2), which is a `let` around a struct. Our code
previously would only handle these for arrays, and not for arrays within
other compound types. Also, Z3 may use multiple bindings (observable in
the same regression test), which we did not cater for at all.

This commit promotes `let` parsing to the top level, expanding bindings
as needed.

tautschnig committed:
Mark tests no longer failing with SMT back-ends as such

These were fixed via some prior commit.

kroening commented:
I think I'd prefer to retain the invariant that the SMT-LIB datatype has the same number of components as the struct. Perhaps use a bit-vector of width 1, and then skip that when flattening?

martin-cs reviewed (approved):
Seems legit.

tautschnig committed:
SMT back-end: implement float to bitvector conversion

Implement what the comment said by creating an appropriate symbol via
`find_symbols`.

tautschnig committed:
SMT back-end: flatten2bv must recurse into arrays and structs

The previous code would only work when array elements/struct members
were bitvectors already. This is not true for nested compound types.

tautschnig committed:
SMT back-end: support unflattening arrays

When data types are used, array-typed members may occur within structs.
Calls to `unflatten` (as is needed for member-of-union expressions) must
then restore array-typed elements and not just bitvectors for struct
constructors.

tautschnig committed:
Try to keep struct and SMTLIB data types in sync

tautschnig committed:
Revert "Try to keep struct and SMTLIB data types in sync"

This reverts commit 0ba0d3200b643dd939b10b92f1258ce9418a4d75.

tautschnig commented:
> I think I'd prefer to retain the invariant that the SMT-LIB datatype has the same number of components as the struct. Perhaps use a bit-vector of width 1, and then skip that when flattening?

I tried this, but at best it makes the code very complicated: it kind of works for structs, but unions containing structs of zero bitwidth require a lot of extra checks. That's because we always flatten unions, so we do expect that the bitwidth is non-zero. I've added a commit ("Try to keep struct and SMTLIB data types in sync") to document my attempt, but it's not actually fully working on the example provided in #7442 (unlike the code without this change, which does work fine on that example).
